### ISSUE TITLE
-Wpadded does nothing when targeting windows

### ISSUE THREAD

**Issue by henke96** [llvm/llvm-project#58100]
test.c:
```c
struct {
    long a;
    char b;
} c;
```
If you compile it with `clang -target x86_64-unknown-linux-elf -Wpadded -c test.c` you get the warning `test.c:1:1: warning: padding size of 'struct (unnamed at test.c:1:1)' with 7 bytes to alignment boundary [-Wpadded]` as expected.

However, if you target MS abi with `clang -target x86_64-unknown-windows-coff -Wpadded -c test.c` there is no warning at all. Note that it happens in real code too, where the struct is actually used. This is just a minimal example.

**Comment by alvinhochun** [comment]
This diag (`diag::warn_unnecessary_packed`) is emitted in `ItaniumRecordLayoutBuilder::FinishLayout`, which is indirectly called from `ASTContext::getASTRecordLayout`. There is a separate code path for MS ABI layout which does not have similar code to handle this diag. (`diag::warn_padded_struct_size` seems to suffer from the same fate.)

https://github.com/llvm/llvm-project/blob/8193f0d1d26b7b4d602dcd5581b40382e4b6cf36/clang/lib/AST/RecordLayoutBuilder.cpp#L2180-L2203

If anyone wants to implement this diag for MS ABI, I guess it should be added in `MicrosoftRecordLayoutBuilder::finalizeLayout`. https://github.com/llvm/llvm-project/blob/8193f0d1d26b7b4d602dcd5581b40382e4b6cf36/clang/lib/AST/RecordLayoutBuilder.cpp#L3150

**Comment by henke96** [comment]
Yeah, i noticed that the warning does work if i target "x86_64-unknown-windows-gnu" instead. So i guess it's only a problem for MSVC abi.

**Comment by cor3ntin** [comment]
@ojhunt
